### ISSUE TITLE
Altivec tests

### ISSUE THREAD

**Issue by gibbed** [benvanik/xenia#391]
- [ ] dst
 - [ ] dstst
 - [ ] dss
 - [ ] lvebx
 - [ ] lvehx
 - [ ] lvewx
 - [ ] lvewx128
 - [ ] lvsl
 - [ ] lvsl128
 - [ ] lvsr
 - [ ] lvsr128
 - [ ] lvx
 - [ ] lvx128
 - [ ] lvxl
 - [ ] lvxl128
 - [ ] stvebx
 - [ ] stvehx
 - [ ] stvewx
 - [ ] stvewx128
 - [ ] stvx
 - [ ] stvx128
 - [ ] stvxl
 - [ ] stvxl128
 - [ ] lvlx
 - [ ] lvlx128
 - [ ] lvlxl
 - [ ] lvlxl128
 - [ ] lvrx
 - [ ] lvrx128
 - [ ] lvrxl
 - [ ] lvrxl128
 - [ ] stvlx
 - [ ] stvlx128
 - [ ] stvlxl
 - [ ] stvlxl128
 - [ ] stvrx
 - [ ] stvrx128
 - [ ] stvrxl
 - [ ] stvrxl128
 - [ ] mfvscr
 - [ ] mtvscr
 - [ ] vaddcuw
 - [ ] vaddfp
 - [ ] vaddfp128
 - [ ] vaddsbs
 - [ ] vaddshs
 - [ ] vaddsws
 - [ ] vaddubm
 - [ ] vaddubs
 - [ ] vadduhm
 - [ ] vadduhs
 - [ ] vadduwm
 - [ ] vadduws
 - [ ] vand
 - [ ] vand128
 - [ ] vandc
 - [ ] vandc128
 - [ ] vavgsb
 - [ ] vavgsh
 - [ ] vavgsw
 - [ ] vavgub
 - [ ] vavguh
 - [ ] vavguw
 - [ ] vcfsx
 - [ ] vcsxwfp128
 - [ ] vcfpsxws128
 - [ ] vcfux
 - [ ] vcuxwfp128
 - [ ] vcfpuxws128
 - [ ] vcmpbfp
 - [ ] vcmpbfp128
 - [ ] vcmpeqfp
 - [ ] vcmpeqfp128
 - [ ] vcmpgefp
 - [ ] vcmpgefp128
 - [ ] vcmpgtfp
 - [ ] vcmpgtfp128
 - [ ] vcmpgtsb
 - [ ] vcmpgtsh
 - [ ] vcmpgtsw
 - [ ] vcmpequb
 - [ ] vcmpgtub
 - [ ] vcmpequh
 - [ ] vcmpgtuh
 - [ ] vcmpequw
 - [ ] vcmpequw128
 - [ ] vcmpgtuw
 - [ ] vctsxs
 - [ ] vctuxs
 - [ ] vexptefp
 - [ ] vexptefp128
 - [ ] vlogefp
 - [ ] vlogefp128
 - [ ] vmaddfp
 - [ ] vmaddfp128
 - [ ] vmaddcfp128
 - [ ] vmaxfp
 - [ ] vmaxfp128
 - [ ] vmaxsb
 - [ ] vmaxsh
 - [ ] vmaxsw
 - [ ] vmaxub
 - [ ] vmaxuh
 - [ ] vmaxuw
 - [ ] vmhaddshs
 - [ ] vmhraddshs
 - [ ] vminfp
 - [ ] vminfp128
 - [ ] vminsb
 - [ ] vminsh
 - [ ] vminsw
 - [ ] vminub
 - [ ] vminuh
 - [ ] vminuw
 - [ ] vmladduhm
 - [ ] vmrghb
 - [ ] vmrghh
 - [ ] vmrghw
 - [ ] vmrghw128
 - [ ] vmrglb
 - [ ] vmrglh
 - [ ] vmrglw
 - [ ] vmrglw128
 - [ ] vmsummbm
 - [ ] vmsumshm
 - [ ] vmsumshs
 - [ ] vmsumubm
 - [ ] vmsumuhm
 - [ ] vmsumuhs
 - [ ] vmsum3fp128
 - [ ] vmsum4fp128
 - [ ] vmulesb
 - [ ] vmulesh
 - [ ] vmuleub
 - [ ] vmuleuh
 - [ ] vmulosb
 - [ ] vmulosh
 - [ ] vmuloub
 - [ ] vmulouh
 - [ ] vmulfp128
 - [ ] vnmsubfp
 - [ ] vnmsubfp128
 - [ ] vnor
 - [ ] vnor128
 - [ ] vor
 - [ ] vor128
 - [ ] vperm
 - [ ] vperm128
 - [ ] vpermwi128
 - [ ] vpkpx
 - [ ] vpkshss
 - [ ] vpkshss128
 - [ ] vpkshus
 - [ ] vpkshus128
 - [ ] vpkswss
 - [ ] vpkswss128
 - [ ] vpkswus
 - [ ] vpkswus128
 - [ ] vpkuhum
 - [ ] vpkuhum128
 - [ ] vpkuhus
 - [ ] vpkuhus128
 - [ ] vpkuwum
 - [ ] vpkuwum128
 - [ ] vpkuwus
 - [ ] vpkuwus128
 - [ ] vpkd3d128
 - [ ] vrefp
 - [ ] vrefp128
 - [ ] vrfim
 - [ ] vrfim128
 - [ ] vrfin
 - [ ] vrfin128
 - [ ] vrfip
 - [ ] vrfip128
 - [ ] vrfiz
 - [ ] vrfiz128
 - [ ] vrlb
 - [ ] vrlh
 - [ ] vrlw
 - [ ] vrlw128
 - [ ] vrlimi128
 - [ ] vrsqrtefp
 - [ ] vrsqrtefp128
 - [ ] vsel
 - [ ] vsel128
 - [ ] vsl
 - [ ] vslb
 - [ ] vslh
 - [ ] vslo
 - [ ] vslo128
 - [ ] vslw
 - [ ] vslw128
 - [ ] vsldoi
 - [ ] vsldoi128
 - [ ] vspltb
 - [ ] vsplth
 - [ ] vspltw
 - [ ] vspltw128
 - [ ] vspltisb
 - [ ] vspltish
 - [ ] vspltisw
 - [ ] vspltisw128
 - [ ] vsr
 - [ ] vsrab
 - [ ] vsrah
 - [ ] vsraw
 - [ ] vsraw128
 - [ ] vsrb
 - [ ] vsrh
 - [ ] vsro
 - [ ] vsro128
 - [ ] vsrw
 - [ ] vsrw128
 - [ ] vsubcuw
 - [ ] vsubfp
 - [ ] vsubfp128
 - [ ] vsubsbs
 - [ ] vsubshs
 - [ ] vsubsws
 - [ ] vsububm
 - [ ] vsubuhm
 - [ ] vsubuwm
 - [ ] vsububs
 - [ ] vsubuhs
 - [ ] vsubuws
 - [ ] vsumsws
 - [ ] vsum2sws
 - [ ] vsum4sbs
 - [ ] vsum4shs
 - [ ] vsum4ubs
 - [ ] vupkhpx
 - [ ] vupkhsb
 - [ ] vupkhsb128
 - [ ] vupkhsh
 - [ ] vupklpx
 - [ ] vupklsb
 - [ ] vupklsb128
 - [ ] vupklsh
 - [ ] vupkd3d128
 - [ ] vxor
 - [ ] vxor128

**Comment by DrChat** [comment]
These instructions are unsupported on the Xenon CPU ([Source](http://gcc.1065356.n5.nabble.com/Implementing-VMX128-AltiVec-Variant-td690065.html)):
vmuloub, vmulouh, vmulosb, vmulosh, vmuleub, vmuleuh, vmulesb, vmulesh, 
vsum4ubs, vsum4shs, vsum2sws, vsum4sbs, vsumsws, vmhaddshs, vmhraddshs, 
vmladduhm, vmsumubm, vmsummbm, vmsumuhm, vmsumuhs, vmsumshm, vmsumshs

**Comment by vgturtle127** [comment]
Are these tests going to be open-sourced, or are you just testing all the stuff you implemented in issue form so you know what you have done?

Can you just explain what these are? I understand that some are and aren't implemented, but other than that I guess I am not understanding... :blush:

**Comment by DrChat** [comment]
These tests are implemented [here](https://github.com/benvanik/xenia/tree/master/src/xenia/cpu/frontend/testing). They test specific CPU instructions, which verifies that both the frontend (PPC parsing) as well as the x64 backend are working correctly.

If you want to implement some AltiVec tests, take a look at [this](https://cache.freescale.com/files/32bit/doc/ref_manual/ALTIVECPEM.pdf).

**Comment by vgturtle127** [comment]
Are the ALU/FPU tests in there as well?

Also, will there eventually be general tests, like RGB triangles and what not?

**Comment by DrChat** [comment]
Yes, those tests are there as well. I don't know about GPU tests, you'd
have to ask @benvanik.

**Comment by vgturtle127** [comment]
Alright. @benvanik Are we going to write GPU tests? :smile: 

Do you guys know anything about the GPU? I mean, it's a custom ATI chip, right? But are there any documents or anything?
